<a href="https://colab.research.google.com/github/ahmadd4062/flyrank-ml-internship/blob/main/work/notebooks/w03_data_contract.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ML-04 — Search Intelligence Data Contract

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

In [1]:
from google.colab import userdata

HF_TOKEN = userdata.get("HF_TOKEN")

print("Token loaded:", HF_TOKEN is not None)

Token loaded: True


In [2]:
from huggingface_hub import snapshot_download
from google.colab import userdata

repo_path = snapshot_download(
    repo_id="FlyRank/internship-warehouse",
    repo_type="dataset",
    token=userdata.get("HF_TOKEN")
)

print(repo_path)

Reconstructing (incomplete total...): |          |  0.00B /  0.00B            

Fetching 24 files:   0%|          | 0/24 [00:00<?, ?it/s]

/root/.cache/huggingface/hub/datasets--FlyRank--internship-warehouse/snapshots/50cbf7c3909d07be4d1b5906b4d09e882e5acbf2


In [3]:
import os

for root, dirs, files in os.walk(repo_path):
    for f in files:
        print(os.path.join(root, f))

/root/.cache/huggingface/hub/datasets--FlyRank--internship-warehouse/snapshots/50cbf7c3909d07be4d1b5906b4d09e882e5acbf2/fact_content_query_90d.parquet
/root/.cache/huggingface/hub/datasets--FlyRank--internship-warehouse/snapshots/50cbf7c3909d07be4d1b5906b4d09e882e5acbf2/.gitattributes
/root/.cache/huggingface/hub/datasets--FlyRank--internship-warehouse/snapshots/50cbf7c3909d07be4d1b5906b4d09e882e5acbf2/fact_content_daily_performance_sample.parquet
/root/.cache/huggingface/hub/datasets--FlyRank--internship-warehouse/snapshots/50cbf7c3909d07be4d1b5906b4d09e882e5acbf2/dim_content.parquet
/root/.cache/huggingface/hub/datasets--FlyRank--internship-warehouse/snapshots/50cbf7c3909d07be4d1b5906b4d09e882e5acbf2/README.md
/root/.cache/huggingface/hub/datasets--FlyRank--internship-warehouse/snapshots/50cbf7c3909d07be4d1b5906b4d09e882e5acbf2/dim_clients.parquet
/root/.cache/huggingface/hub/datasets--FlyRank--internship-warehouse/snapshots/50cbf7c3909d07be4d1b5906b4d09e882e5acbf2/fact_content_daily

In [4]:
import pandas as pd
import os

march_path = os.path.join(
    repo_path,
    "fact_content_daily_performance",
    "month=2026-03",
    "data_0.parquet"
)

df = pd.read_parquet(march_path)

print(df.shape)
df.head()

(9841378, 30)


,report_date,client_hash_id,content_hash_id,client_has_gsc,client_has_ga4,gsc_data_available,ga4_data_available,gsc_impressions,gsc_clicks,gsc_sum_position,...,sessions_paid,sessions_ai,ai_chatgpt,ai_perplexity,ai_gemini,ai_copilot,ai_claude,ai_meta,ai_other,scroll_events
0,2026-03-01,client_73cda7b4e4f265ea,content_b7e512995f79d5a6,True,False,True,None,20,0,67,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,2026-03-01,client_73cda7b4e4f265ea,content_05597932fe4da067,True,False,True,None,1,0,0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,2026-03-01,client_73cda7b4e4f265ea,content_7a105f548d9c6916,True,False,True,None,125,1,616,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,2026-03-01,client_73cda7b4e4f265ea,content_905aa32a0230694e,True,False,True,None,7,0,28,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,2026-03-01,client_73cda7b4e4f265ea,content_a3ea9792f793ec72,True,False,True,None,11,0,25,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [5]:
print(df.columns.tolist())

['report_date', 'client_hash_id', 'content_hash_id', 'client_has_gsc', 'client_has_ga4', 'gsc_data_available', 'ga4_data_available', 'gsc_impressions', 'gsc_clicks', 'gsc_sum_position', 'gsc_avg_position', 'ga4_pageviews', 'ga4_sessions', 'ga4_users', 'ga4_engaged_sessions', 'ga4_total_engagement_sec', 'sessions_organic', 'sessions_direct', 'sessions_referral', 'sessions_social', 'sessions_paid', 'sessions_ai', 'ai_chatgpt', 'ai_perplexity', 'ai_gemini', 'ai_copilot', 'ai_claude', 'ai_meta', 'ai_other', 'scroll_events']


In [13]:
feature_cols = [
    "gsc_impressions",
    "gsc_clicks",
    "gsc_avg_position",
    "ga4_pageviews",
    "sessions_organic"
]

df[feature_cols].describe()

,gsc_impressions,gsc_clicks,gsc_avg_position,ga4_pageviews,sessions_organic
count,9.841378e+06,9.841378e+06,3.611061e+06,6.822637e+06,6.822637e+06
mean,2.851812e+01,8.350782e-02,1.582665e+01,2.176367e-01,8.580377e-02
std,1.559266e+02,7.814341e-01,1.985603e+01,2.142851e+00,8.754357e-01
min,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00
25%,0.000000e+00,0.000000e+00,3.742120e+00,0.000000e+00,0.000000e+00
50%,0.000000e+00,0.000000e+00,7.500000e+00,0.000000e+00,0.000000e+00
75%,6.000000e+00,0.000000e+00,2.020000e+01,0.000000e+00,0.000000e+00
max,4.008400e+04,2.740000e+02,4.980000e+02,8.750000e+02,4.050000e+02


In [14]:
df[feature_cols].head()

,gsc_impressions,gsc_clicks,gsc_avg_position,ga4_pageviews,sessions_organic
0,20,0,3.350000,NaN,NaN
1,1,0,0.000000,NaN,NaN
2,125,1,4.928000,NaN,NaN
3,7,0,4.000000,NaN,NaN
4,11,0,2.272727,NaN,NaN


## 1. Unit of analysis + time window

*One row = one what, over which dates? State it, then verify it below.*

## Unit of Analysis + Time Window

- **One row** represents the performance of one content page (`content_hash_id`) for one client (`client_hash_id`) on one reporting date (`report_date`).
- **Time window:** March 2026 (`month=2026-03`) from the `fact_content_daily_performance` table.
- This time window is used because it is a mid-panel month and avoids using the final month as development data.

## 2. Fields: feature / label / context / excluded

*Sort every field you plan to touch into these four buckets. Excluded needs a why.*

## Fields

### Features
- gsc_impressions
- gsc_clicks
- gsc_avg_position
- client_has_gsc
- gsc_data_available

These features are known at the decision time because they are historical search performance signals available before deciding whether a page should be reviewed.

### Label / Proxy
Refresh opportunity priority (proxy).

There is no direct "needs refresh" label in the dataset, so the model uses observed search performance as a proxy for prioritizing pages for review.

### Context
- report_date
- client_hash_id
- content_hash_id

These identify when the observation occurred and which client and content page it belongs to.

### Excluded
GA4 metrics (pageviews, sessions, users, engagement metrics) are excluded because they are unavailable for many clients, resulting in many missing values. AI referral fields are also excluded because they are not consistently available across all observations.

## 3. Verify it with queries (grain, counts, missing values, windows)

*Every claim above gets a query cell here. A contract claim without a query next to it is a guess.*


In [15]:
# Query 1: Grain (What does one row represent)
print("Total rows:", len(df))

print("\nUnique report dates:", df["report_date"].nunique())
print("Unique content pages:", df["content_hash_id"].nunique())
print("Unique clients:", df["client_hash_id"].nunique())

duplicates = df.duplicated(
    subset=["report_date", "client_hash_id", "content_hash_id"]
).sum()

print("\nDuplicate (date, client, content) rows:", duplicates)

Total rows: 9841378

Unique report dates: 31
Unique content pages: 331437
Unique clients: 55

Duplicate (date, client, content) rows: 0


The query confirms the grain of the dataset. There are no duplicate combinations of report date, client, and content page, supporting the claim that one row represents one content page for one client on one reporting date.

In [16]:
# Query 2: Row count + date span
print("Rows:", len(df))
print("Start date:", df["report_date"].min())
print("End date:", df["report_date"].max())

Rows: 9841378
Start date: 2026-03-01
End date: 2026-03-31


The selected partition contains 9,841,378 rows covering the period from 2026-03-01 to 2026-03-31.

In [17]:
# Query 3: Availability (IS TRUE equivalent)
available = df[
    (df["gsc_data_available"] == True)
]

print("Rows with GSC data available:", len(available))

Rows with GSC data available: 3611061


Filtering on `gsc_data_available == True` leaves 3,611,061 rows. This verifies that not every observation contains Search Console data, so feature availability must be considered during analysis.

## 4. Data limits

*What can this data never tell you? Unbalanced history, GSC-only early rows, window overlaps.*

## Data Limits

This dataset contains historical search-performance observations rather than direct labels indicating whether refreshing a page improved future performance. Therefore, the model provides decision support rather than proof that a refresh will increase traffic or rankings.

Some fields, particularly GA4 metrics, are unavailable for clients without Google Analytics integration, resulting in missing values. Additionally, this analysis uses a single monthly partition (March 2026), so the observations represent that time window only and should not be interpreted as long-term trends.

## Self-check

Before you submit, confirm each line honestly:

- [ yes] Every section above is filled — markdown thinking AND the code that backs it
- [yes ] The notebook runs top to bottom with no errors (Runtime → Run all)
- [ yes] No client names, URLs, or private queries anywhere
- [ yes] My claims use careful words: observed, measured, directional, decision-support
- [ yes] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.